Deprecate sorting in reindex

henningandersen opened the pull request:
Reindex sort never gave a guarantee about the order of documents being
indexed into the destination, though it could give a sense of locality
of source data.

It prevents us from doing resilient reindex and other optimizations and
it has therefore been deprecated.

Related to #47567

The deprecation notice in [ILM documentations use of sorted reindex](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/ilm-reindexing-into-rollover.html) may need further refinement once we move to removing sort.

henningandersen committed:
Deprecate sorting in reindex

Reindex sort never gave a guarantee about the order of documents being
indexed into the destination, though it could give a sense of locality
of source data.

It prevents us from doing resilient reindex and other optimizations and
it has therefore been deprecated.

Related to #47567

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Reindex)

henningandersen committed:
Fixed test failure

Had to skip testing this prior to 7.6 in order to add the warning.

henningandersen committed:
Improve documentation layout

henningandersen committed:
Improve documentation layout

henningandersen committed:
Merge remote-tracking branch 'origin/master' into deprecate_reindex_sort

henningandersen committed:
Merge remote-tracking branch 'origin/master' into deprecate_reindex_sort

Tim-Brooks reviewed (approved):
LGTM

jrodewig commented on docs/reference/docs/reindex.asciidoc:
```diff
@@ -478,6 +478,10 @@ which defaults to a maximum size of 100 MB.
+deprecated[7.6, sort in reindex is deprecated] *NOTE: Sorting in reindex is deprecated.*
+Sorting in reindex was never guaranteed to index
+documents in order and prevents future evolution of reindex such as resilience and performance
+improvements. If used in combination with `max_docs`, consider using a query filter instead.
```
I feel like the `deprecated` macro and the bold note are a bit redundant.

I'd place this content into a block `deprecated` macro like below.

```
`sort`:::
+
--
(Optional, list) A comma-separated list of `<field>:<direction>` pairs to sort by before indexing.
Use in conjunction with `max_docs` to control what documents are reindexed.

deprecated::[7.6,Sort in reindex is deprecated. Sorting in reindex was never guaranteed to index documents in order and prevents reindex such as resilience and performance improvements. If used in combination with `max_docs`, consider using a query filter instead.]
--
```

henningandersen replied:
I tried this now, but I get an error saying:

```
element listitem: validity error : Element listitem content does not follow the DTD
```
which is why I originally went for the simpler solution. My asciidoc skills are not super-advanced, so maybe there is a trick to make this work?

jrodewig replied:
I was able to get this working on my end without that error.

<img width="725" alt="Screen Shot 2019-11-27 at 3 43 43 PM" src="https://user-images.githubusercontent.com/40268737/69758153-be782700-112c-11ea-86f3-e5d2c636036b.png">

Here's the commit on my fork if that helps:
https://github.com/jrodewig/elasticsearch/commit/7111829c9d08cba24af5dc776c53e958a7490a58

Feel free to cherry-pick it or copy/paste the contents.

There are a few tricks:
* The delimiter setup is needed.
* The `deprecated` text must be on a single line.

jrodewig replied:
What you have is workable though. If you're not able to get the streamlined deprecated note working, I wouldn't hold up this PR. We can reformat it as a separate issue.

henningandersen replied:
Thanks so much @jrodewig , worked out fine on one line (had to also escape the comma), see: aa2a7e04354b4389ac886ee42f1ecbf634d5fb8a

jrodewig commented on docs/reference/docs/reindex.asciidoc:
```diff
@@ -478,6 +478,10 @@ which defaults to a maximum size of 100 MB.
 Use in conjunction with `max_docs` to control what documents are reindexed.
+deprecated[7.6, sort in reindex is deprecated] *NOTE: Sorting in reindex is deprecated.*
+Sorting in reindex was never guaranteed to index
+documents in order and prevents future evolution of reindex such as resilience and performance
```
I'd go with "further" rather than "future"
```suggestion
documents in order and prevents further development of reindex, such as resilience and performance
```

jrodewig reviewed (commented):
Left a suggestion re: streamlining the `deprecated` note. Also left a minor editorial comment.

Otherwise LGTM.

jrodewig commented on docs/reference/docs/reindex.asciidoc:
```diff
@@ -837,8 +818,8 @@ POST _reindex
 
-<1> `_reindex` defaults to sorting by `_doc` so `random_score` will not have any
-effect unless you override the sort to `_score`.
+<1> you may need to adjust the `min_score` depending on the relative amount of
```
```suggestion
<1> You may need to adjust the `min_score` depending on the relative amount of
```

henningandersen committed:
Documentation fixes

Streamline deprecation notice and better text.

Add deprecation notice to ILM reindex doc.

jrodewig reviewed (approved):
LGTM. I can address reformatting the deprecated note as a separate issue.

henningandersen commented:
Thanks for reviewing @jrodewig and @tbrooks8.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into deprecate_reindex_sort

henningandersen commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/packaging-sample-matrix
